### ISSUE TITLE
Changing name of ExpoKit app breaks iOS build / ExpoKit depends on App.json name in strange ways

### ISSUE THREAD

**Issue by llamaluvr** [expo/expo#3586]
### Environment

```
Expo CLI 2.11.3 environment info:
    System:
      OS: macOS 10.14.2
      Shell: 3.2.57 - /bin/bash
    Binaries:
      Node: 10.14.2 - ~/.nvm/versions/node/v10.14.2/bin/node
      Yarn: 1.13.0 - /usr/local/bin/yarn
      npm: 6.4.1 - ~/.nvm/versions/node/v10.14.2/bin/npm
      Watchman: 4.9.0 - /usr/local/bin/watchman
    IDEs:
      Android Studio: 3.2 AI-181.5540.7.32.5056338
      Xcode: 10.1/10B61 - /usr/bin/xcodebuild
    npmGlobalPackages:
      expo-cli: 2.11.3
```

Targets: ExpoKit, iOS, Android

### Steps to Reproduce

1. Create an ExpoKit app with `expo init` (advanced workflow)
2. Restore your cocoapods (`cd ios; pod install`)
3. Change the name value in app.json
4. Build the iOS app in Xcode
5. Build the app in Android

#### iOS
The app builds.

#### Android
The app builds AND has no artifacts of the name change because, once you're ejected, the name of your app should depend on `app_name` in strings.xml, and not anything in app.json.

#### iOS
The build fails with a "Shell Script invocation error". Error details:
```
Preparing iOS build at /Users/keith/Documents/GitHub/nameofapp/ios...
[10:44:56] ENOENT: no such file or directory, open '/Users/keith/Documents/GitHub/nameofapp/ios/the-other-name-of-my-app/Supporting/EXBuildConstants.json'
[10:44:56] Set EXPO_DEBUG=true in your env to view the stack trace.
Command /bin/sh failed with exit code 1
```

If I put the name in app.json back to what it originally was, the build works again.

#### Android
The app builds. The name of the app on the homescreen, launcher, etc, will match whatever is in `@app_name`. But, if you press the app switching key to get to the recent apps list, your app will have the name from `expo.name` in app.json at the top of its card. This is because the name for the android task is set based on the manifest name. See code at: https://github.com/expo/expo/search?q=setTaskDescription&unscoped_q=setTaskDescription

### Reproducible Demo

https://github.com/llamaluvr/test-expokit-name-change-error

### Description

When an Expo project is ejected, the name of the project folder in the ios folder gets set to a slug-ified version of the `expo.name` field value at the time of ejecting. At first glance, this isn't that much of a problem: once you eject, the name of the app is generally set by values in Info.plist or strings.xml. It does throw one off a bit that changing such a likely-to-change value could break the build (maybe this folder name could just be static, like "expokit-app", or based off of some less-likely to change value than the display name of the app).

However, when it comes to the Android side, this lack of ability to change the name puts one in a catch-22, as the name in app.json sets the name for the task in the recent apps view. So, if I change the name of my app, either my iOS build breaks, or I have to live with the old name showing up in the Android recent apps view (or I have to work around this in native code somehow).

**Comment by tcdavis** [comment]
Thanks @llamaluvr!
Unfortunately this is one of many instances of strange behaviour when working with ExpoKit which we're attempting to address in one fell swoop.
You can read about what the plans here (https://blog.expo.io/expokit-2019-1e5cb02106f8) but we should have something ready for you to use in the coming days.

**Comment by N1ghtly** [comment]
Hi @tcdavis, did you guys find a solution for this? Our team is running into the same problem.

**Comment by llamaluvr** [comment]
@N1ghtly I think what @tcdavis is saying is that, with unimodules, all the tight-coupling between the Expo libraries and the native projects will go away, so an ejected project will not have weird code that breaks based on Expo manifest changes, since it will not depend on an Expo manifest.

I don't think this necessarily fixes the issue for a lot of us right now, because I don't want to ditch ExpoKit and lose OTA updates, which are not available as a unimodule. However, the latest post on the subject (https://blog.expo.io/you-can-now-use-expo-apis-in-any-react-native-app-7c3a93041331) says the plan is to make OTA updates a unimodule. At that point, I think ExpoKit in its current form is obsolete for most of us.

If you need to change your project name across the board, I think you could work around the issue by changing the name in your app.json, then changing the name of the folder/ project files in the ios folder to match the slugified name, and then, of course, change the name in any of the plist/ strings files that affect other name references on iOS or Android.

In our case, we have a white label build process to output multiple apps based on a single codebase, so it's not quite that straightforward. I did find that https://github.com/jwarby/react-native-android-taskdescription works well for changing the name in the recent apps view once the JavaScript is loaded. However, it still shows the app.json name in recent apps while the app is loading/ downloading any OTA updates. It appears to be difficult to change that without forking the ExpoKit node module, as that name setting occurs asynchronously.

**Comment by tcdavis** [comment]
Thanks for translating me :)
Yes, that's mostly spot on, although I think there will still be some dependency on the manifest as that's necessary for the updates API.
We're focused on bringing the rest of the APIs (like updates and background location) to unimodules before recommending existing projects transition over.  Thanks for your patience during this period, but we think this is the best way to get everyone to a happy place.

**Comment by N1ghtly** [comment]
I figured out a temporary fix for this guys :)

After the publish, replace the name in the android/app/src/main/assets/shell-app-manifest.json to whatever you need. It doesn't seem to break anything, but the name will be used in the recent apps screen.

**Comment by dolsem** [comment]
I was able to fix this issue by simply symlinking my iOS project directory with the slugged version of the name as it appears in app.json.

**Comment by oshimayoan** [comment]
Hey @dolsem, may I know how do you that?

**Comment by dolsem** [comment]
@oshimayoan this solution works on Linux and Mac OS, probably not an option on Windows. Imagine you have an app called `MyApp`. Expo would use its slugged version (in this case `myapp`) when generating iOS project. Imagine you change the name to `MyNewApp` afterwards. What I then did is use grep to find every occurrence of `myapp` in iOS project files and replace it with `MyNewApp`. (a bit tedious, but solves most problems). Then I encountered the same issue that is reported here. What I then did was simply go to iOS folder and run `ln -s ./MyNewApp ./myapp`. The issue went away and I was able to build it successfully. I also added `ios/myapp` to .gitignore, so that it doesn't show up in git.
